File: dump-ble.m
#import <CoreBluetooth/CoreBluetooth.h>
#import <Foundation/Foundation.h>

static NSFileHandle *logFile = nil;

static void Log(NSString *format, ...) {
    va_list args;
    va_start(args, format);
    NSString *msg = [[NSString alloc] initWithFormat:format arguments:args];
    va_end(args);
    NSLog(@"%@", msg);
    if (logFile) {
        [logFile writeData:[[msg stringByAppendingString:@"\n"] dataUsingEncoding:NSUTF8StringEncoding]];
    }
}

static NSString *Hex(NSData *data) {
    const unsigned char *bytes = data.bytes;
    NSMutableArray<NSString *> *parts = [NSMutableArray arrayWithCapacity:data.length];
    for (NSUInteger i = 0; i < data.length; i++) {
        [parts addObject:[NSString stringWithFormat:@"%02x", bytes[i]]];
    }
    return [parts componentsJoinedByString:@" "];
}

static NSString *Ascii(NSData *data) {
    const unsigned char *bytes = data.bytes;
    NSMutableString *result = [NSMutableString stringWithCapacity:data.length];
    for (NSUInteger i = 0; i < data.length; i++) {
        unsigned char byte = bytes[i];
        [result appendFormat:@"%c", byte >= 0x20 && byte <= 0x7e ? byte : '.'];
    }
    return result;
}

static void PrintData(NSString *source, NSData *data) {
    if (!data || data.length == 0) return;
    Log(@"DATA %@ len=%lu hex=%@ ascii=%@", source, (unsigned long)data.length, Hex(data), Ascii(data));
}

static NSString *ShortProps(CBCharacteristicProperties p) {
    NSMutableArray *a = [NSMutableArray array];
    if (p & CBCharacteristicPropertyRead) [a addObject:@"R"];
    if (p & CBCharacteristicPropertyWrite) [a addObject:@"W"];
    if (p & CBCharacteristicPropertyWriteWithoutResponse) [a addObject:@"WNR"];
    if (p & CBCharacteristicPropertyNotify) [a addObject:@"N"];
    if (p & CBCharacteristicPropertyIndicate) [a addObject:@"I"];
    if (p & CBCharacteristicPropertyNotifyEncryptionRequired) [a addObject:@"Nenc"];
    if (p & CBCharacteristicPropertyIndicateEncryptionRequired) [a addObject:@"Ienc"];
    return [a componentsJoinedByString:@","];
}

@interface BleDumper : NSObject <CBCentralManagerDelegate, CBPeripheralDelegate>
@property(nonatomic, strong) CBCentralManager *central;
@property(nonatomic, strong) CBPeripheral *peripheral;
@property(nonatomic, strong) NSMutableDictionary<NSString *, CBCharacteristic *> *allChars;
@property(nonatomic, assign) BOOL didConnect;
@property(nonatomic, assign) BOOL didDiscover;
@property(nonatomic, assign) int pairAttempts;
@end

@implementation BleDumper

- (instancetype)init {
    self = [super init];
    _allChars = [NSMutableDictionary dictionary];
    _central = [[CBCentralManager alloc] initWithDelegate:self queue:dispatch_get_main_queue()];
    return self;
}

- (void)centralManagerDidUpdateState:(CBCentralManager *)central {
    Log(@"STATE: %ld", (long)central.state);
    if (central.state != CBManagerStatePoweredOn) return;
    Log(@"SCANNING for 'nanokey'...");
    [central scanForPeripheralsWithServices:nil options:@{CBCentralManagerScanOptionAllowDuplicatesKey: @YES}];
}

- (void)centralManager:(CBCentralManager *)central didDiscoverPeripheral:(CBPeripheral *)peripheral advertisementData:(NSDictionary<NSString *,id> *)adData RSSI:(NSNumber *)rssi {
    NSString *name = adData[CBAdvertisementDataLocalNameKey] ?: peripheral.name ?: @"";
    if ([name.lowercaseString rangeOfString:@"nanokey"].location == NSNotFound) return;

    Log(@"FOUND: %@ (id=%@ rssi=%@)", name, peripheral.identifier.UUIDString, rssi);
    self.peripheral = peripheral;
    peripheral.delegate = self;
    [central stopScan];

    Log(@"CONNECTING...");
    [central connectPeripheral:peripheral options:@{
        CBConnectPeripheralOptionNotifyOnConnectionKey: @YES,
        CBConnectPeripheralOptionNotifyOnDisconnectionKey: @YES,
        CBConnectPeripheralOptionNotifyOnNotificationKey: @YES,
    }];
}

- (void)centralManager:(CBCentralManager *)central didConnectPeripheral:(CBPeripheral *)peripheral {
    Log(@"CONNECTED. Starting discovery in 1s...");
    self.didConnect = YES;
    dispatch_after(dispatch_time(DISPATCH_TIME_NOW, 1 * NSEC_PER_SEC), dispatch_get_main_queue(), ^{
        [peripheral discoverServices:nil];
    });
}

- (void)centralManager:(CBCentralManager *)central didFailToConnectPeripheral:(CBPeripheral *)peripheral error:(NSError *)error {
    Log(@"CONNECT FAIL: %@", error);
    exit(1);
}

- (void)centralManager:(CBCentralManager *)central didDisconnectPeripheral:(CBPeripheral *)peripheral error:(NSError *)error {
    Log(@"DISCONNECTED: %@ (code=%ld)", error.localizedDescription ?: @"no error", (long)error.code);
    if (error && (error.code == CBErrorConnectionTimeout || error.code == CBErrorPeripheralDisconnected)) {
        Log(@"Device requires pairing to stay connected. Check Bluetooth system dialog.");
    }
    exit(0);
}

- (void)peripheral:(CBPeripheral *)peripheral didDiscoverServices:(NSError *)error {
    if (error) { Log(@"SERVICE ERR: %@", error); exit(1); }
    Log(@"SERVICES: %lu found", (unsigned long)peripheral.services.count);
    for (CBService *s in peripheral.services) {
        Log(@"  svc: %@", s.UUID.UUIDString);
        [peripheral discoverCharacteristics:nil forService:s];
    }
}

- (void)peripheral:(CBPeripheral *)peripheral didDiscoverCharacteristicsForService:(CBService *)service error:(NSError *)error {
    if (error) { Log(@"CHAR ERR for %@: %@", service.UUID.UUIDString, error); return; }
    Log(@"CHARS for %@: %lu found", service.UUID.UUIDString, (unsigned long)service.characteristics.count);
    for (CBCharacteristic *c in service.characteristics) {
        NSString *key = [NSString stringWithFormat:@"%@/%@", service.UUID.UUIDString, c.UUID.UUIDString];
        self.allChars[key] = c;
        Log(@"  char: %@ [%@]", c.UUID.UUIDString, ShortProps(c.properties));

        if (c.properties & CBCharacteristicPropertyRead) {
            Log(@"  -> reading %@ (may trigger pairing dialog!)", key);
            [peripheral readValueForCharacteristic:c];
        }
    }

    dispatch_after(dispatch_time(DISPATCH_TIME_NOW, 3 * NSEC_PER_SEC), dispatch_get_main_queue(), ^{
        [self enableNotifications];
    });
}

- (void)enableNotifications {
    Log(@"Enabling notifications on appropriate characteristics...");
    for (NSString *key in self.allChars) {
        CBCharacteristic *c = self.allChars[key];
        if ((c.properties & CBCharacteristicPropertyNotify) || (c.properties & CBCharacteristicPropertyIndicate)) {
            Log(@"  -> enabling notify on %@...", key);
            [self.peripheral setNotifyValue:YES forCharacteristic:c];
        }
    }
}

- (void)peripheral:(CBPeripheral *)peripheral didUpdateValueForCharacteristic:(CBCharacteristic *)characteristic error:(NSError *)error {
    NSString *key = [NSString stringWithFormat:@"%@/%@", characteristic.service.UUID.UUIDString, characteristic.UUID.UUIDString];

    if (error) {
        Log(@"READ ERR %@: %@ (domain=%@ code=%ld)", key, error.localizedDescription, error.domain, (long)error.code);
        if (error.code == CBATTErrorInsufficientAuthentication || error.code == CBATTErrorInsufficientEncryption) {
            Log(@"  -> Pairing REQUIRED! macOS should show pairing dialog.");
            Log(@"  -> Check: System Settings > Bluetooth");
            Log(@"  -> If dialog didn't appear, try: Audio MIDI Setup > MIDI Studio > Bluetooth Configuration");
            self.pairAttempts++;
            if (self.pairAttempts < 5) {
                dispatch_after(dispatch_time(DISPATCH_TIME_NOW, 3 * NSEC_PER_SEC), dispatch_get_main_queue(), ^{
                    Log(@"  -> RETRYING read %@ (pairing should be in progress)...", key);
                    [peripheral readValueForCharacteristic:characteristic];
                });
            }
        }
        return;
    }

    Log(@"READ OK %@", key);
    PrintData(key, characteristic.value);
}

- (void)peripheral:(CBPeripheral *)peripheral didUpdateNotificationStateForCharacteristic:(CBCharacteristic *)characteristic error:(NSError *)error {
    NSString *key = [NSString stringWithFormat:@"%@/%@", characteristic.service.UUID.UUIDString, characteristic.UUID.UUIDString];
    if (error) {
        Log(@"NOTIFY ERR %@: %@ (domain=%@ code=%ld)", key, error.localizedDescription, error.domain, (long)error.code);
        if (error.code == CBATTErrorInsufficientAuthentication || error.code == CBATTErrorInsufficientEncryption) {
            Log(@"  -> Pairing REQUIRED for notifications!");
        }
        return;
    }
    Log(@"NOTIFY %@: %s", key, characteristic.isNotifying ? "ON" : "OFF");
    if (characteristic.isNotifying) {
        Log(@"  -> READY! Press keys/pads/knobs now. Data will appear here.");
    }
}

- (void)peripheral:(CBPeripheral *)peripheral didDiscoverDescriptorsForCharacteristic:(CBCharacteristic *)characteristic error:(NSError *)error {
    for (CBDescriptor *d in characteristic.descriptors ?: @[]) {
        Log(@"  desc: %@", d.UUID.UUIDString);
    }
}

@end

int main(void) {
    @autoreleasepool {
        NSString *logPath = @"/tmp/korg-dump.log";
        [[NSFileManager defaultManager] createFileAtPath:logPath contents:nil attributes:nil];
        logFile = [NSFileHandle fileHandleForWritingAtPath:logPath];
        [logFile truncateFileAtOffset:0];
        setvbuf(stdout, NULL, _IOLBF, 0);

        Log(@"=== KORG BLE DUMP v3 (with pairing support) ===");
        Log(@"Log: %@", logPath);
        printf("=== KORG BLE DUMP v3 ===\n");
        printf("Log: %s\n", [logPath UTF8String]);
        printf("Make sure nanoKEY Studio is ON and in Bluetooth range.\n");
        printf("If pairing dialog appears - ACCEPT IT.\n");
        printf("Ctrl+C to stop.\n\n");

        BleDumper *d = [BleDumper new];
        (void)d;
        [[NSRunLoop mainRunLoop] run];
    }
    return 0;
}

File: dump-ble.swift
import CoreBluetooth
import Foundation

let deviceNameFilter = ProcessInfo.processInfo.environment["DEVICE_NAME"]?.lowercased() ?? "nanokey"
let runLoop = RunLoop.current

func hex(_ data: Data) -> String {
    data.map { String(format: "%02x", $0) }.joined(separator: " ")
}

func ascii(_ data: Data) -> String {
    String(data.map { byte in
        if byte >= 0x20 && byte <= 0x7e { return Character(UnicodeScalar(byte)) }
        return "."
    })
}

func printData(_ source: String, _ data: Data?) {
    guard let data = data, !data.isEmpty else { return }

    let timestamp = ISO8601DateFormatter().string(from: Date())
    print("[\(timestamp)] \(source)")
    print("  len:   \(data.count)")
    print("  hex:   \(hex(data))")
    print("  ascii: \(ascii(data))")
}

final class BleDumper: NSObject, CBCentralManagerDelegate, CBPeripheralDelegate {
    private var central: CBCentralManager!
    private var peripheral: CBPeripheral?

    override init() {
        super.init()
        central = CBCentralManager(delegate: self, queue: nil)
    }

    func centralManagerDidUpdateState(_ central: CBCentralManager) {
        print("Bluetooth state: \(central.state.rawValue)")

        guard central.state == .poweredOn else {
            print("Bluetooth is not powered on/ready: \(central.state)")
            return
        }

        print("Scanning for device name containing: \(deviceNameFilter)")
        central.scanForPeripherals(withServices: nil, options: [CBCentralManagerScanOptionAllowDuplicatesKey: true])
    }

    func centralManager(
        _ central: CBCentralManager,
        didDiscover peripheral: CBPeripheral,
        advertisementData: [String: Any],
        rssi RSSI: NSNumber
    ) {
        let localName = advertisementData[CBAdvertisementDataLocalNameKey] as? String
        let name = localName ?? peripheral.name ?? "<no name>"

        guard name.lowercased().contains(deviceNameFilter) else { return }

        print("Found \(name)")
        print("identifier: \(peripheral.identifier.uuidString)")
        print("rssi: \(RSSI)")

        if let services = advertisementData[CBAdvertisementDataServiceUUIDsKey] as? [CBUUID] {
            print("advertised services: \(services.map { $0.uuidString }.joined(separator: ", "))")
        }

        if let manufacturerData = advertisementData[CBAdvertisementDataManufacturerDataKey] as? Data {
            printData("manufacturer data", manufacturerData)
        }

        self.peripheral = peripheral
        peripheral.delegate = self
        central.stopScan()
        central.connect(peripheral, options: nil)
    }

    func centralManager(_ central: CBCentralManager, didConnect peripheral: CBPeripheral) {
        print("Connected. Discovering services...")
        peripheral.discoverServices(nil)
    }

    func centralManager(_ central: CBCentralManager, didFailToConnect peripheral: CBPeripheral, error: Error?) {
        print("Failed to connect: \(error?.localizedDescription ?? "unknown error")")
        Foundation.exit(1)
    }

    func centralManager(_ central: CBCentralManager, didDisconnectPeripheral peripheral: CBPeripheral, error: Error?) {
        print("Disconnected: \(error?.localizedDescription ?? "no error")")
        Foundation.exit(0)
    }

    func peripheral(_ peripheral: CBPeripheral, didDiscoverServices error: Error?) {
        if let error = error {
            print("Service discovery failed: \(error.localizedDescription)")
            Foundation.exit(1)
        }

        let services = peripheral.services ?? []
        print("Services (\(services.count)):")
        for service in services {
            print("  \(service.uuid.uuidString)")
            peripheral.discoverCharacteristics(nil, for: service)
        }
    }

    func peripheral(_ peripheral: CBPeripheral, didDiscoverCharacteristicsFor service: CBService, error: Error?) {
        if let error = error {
            print("Characteristic discovery failed for \(service.uuid.uuidString): \(error.localizedDescription)")
            return
        }

        let characteristics = service.characteristics ?? []
        print("Characteristics for service \(service.uuid.uuidString) (\(characteristics.count)):")

        for characteristic in characteristics {
            print("  \(characteristic.uuid.uuidString) properties=\(describeProperties(characteristic.properties))")
            peripheral.discoverDescriptors(for: characteristic)

            if characteristic.properties.contains(.read) {
                peripheral.readValue(for: characteristic)
            }

            if characteristic.properties.contains(.notify) || characteristic.properties.contains(.indicate) {
                print("  enabling notify for \(service.uuid.uuidString)/\(characteristic.uuid.uuidString)")
                peripheral.setNotifyValue(true, for: characteristic)
            }
        }
    }

    func peripheral(_ peripheral: CBPeripheral, didDiscoverDescriptorsFor characteristic: CBCharacteristic, error: Error?) {
        if let error = error {
            print("Descriptor discovery failed for \(characteristic.uuid.uuidString): \(error.localizedDescription)")
            return
        }

        for descriptor in characteristic.descriptors ?? [] {
            print("    descriptor \(descriptor.uuid.uuidString) for characteristic \(characteristic.uuid.uuidString)")
        }
    }

    func peripheral(_ peripheral: CBPeripheral, didUpdateValueFor characteristic: CBCharacteristic, error: Error?) {
        if let error = error {
            print("Value update failed for \(characteristic.uuid.uuidString): \(error.localizedDescription)")
            return
        }

        let serviceUuid = characteristic.service?.uuid.uuidString ?? "<unknown service>"
        printData("value \(serviceUuid)/\(characteristic.uuid.uuidString)", characteristic.value)
    }

    func peripheral(_ peripheral: CBPeripheral, didUpdateNotificationStateFor characteristic: CBCharacteristic, error: Error?) {
        let serviceUuid = characteristic.service?.uuid.uuidString ?? "<unknown service>"

        if let error = error {
            print("Notify failed \(serviceUuid)/\(characteristic.uuid.uuidString): \(error.localizedDescription)")
            return
        }

        print("Notify state \(serviceUuid)/\(characteristic.uuid.uuidString): \(characteristic.isNotifying)")
    }

    private func describeProperties(_ properties: CBCharacteristicProperties) -> String {
        var names: [String] = []
        if properties.contains(.broadcast) { names.append("broadcast") }
        if properties.contains(.read) { names.append("read") }
        if properties.contains(.writeWithoutResponse) { names.append("writeWithoutResponse") }
        if properties.contains(.write) { names.append("write") }
        if properties.contains(.notify) { names.append("notify") }
        if properties.contains(.indicate) { names.append("indicate") }
        if properties.contains(.authenticatedSignedWrites) { names.append("authenticatedSignedWrites") }
        if properties.contains(.extendedProperties) { names.append("extendedProperties") }
        if properties.contains(.notifyEncryptionRequired) { names.append("notifyEncryptionRequired") }
        if properties.contains(.indicateEncryptionRequired) { names.append("indicateEncryptionRequired") }
        return names.isEmpty ? "-" : names.joined(separator: ",")
    }
}

let dumper = BleDumper()
print("CoreBluetooth BLE dump started. Ctrl+C to stop.")

while runLoop.run(mode: .default, before: Date(timeIntervalSinceNow: 1)) {}

_ = dumper
